logic/atrium-logic-common/src/main/kotlin/ch/tutteli/atrium/logic/impl/DefaultIterableLikeAssertions.kt

```diff
@@ -176,8 +176,13 @@ class DefaultIterableLikeAssertions : IterableLikeAssertions {
     ): Assertion = LazyThreadUnsafeAssertionGroup {
         val list = transformToList(container, converter)
 
-        val duplicates = createIndexAssertions(list) { (_, element) ->
-            list.count { e -> e == element } > 1
+        val duplicates = createIndexAssertions(list) { (index, element) ->
```

Comment:
We can do even better here. We need to refactor a bit tough. @jGleitz would you mind giving @shytnik-igor a few hints? Just if you have time, otherwise I'll do it

Reply:
I can try.

@shytnik-igor, your change does indeed improve performance in some cases, because we stop searching once we have found the element. The previous implementation _always_ required a full pass over the list for every element.

Note, however, that your change is only an improvement if there are duplicates. In the expected case, which is that there are no duplicates, we still perform O(n²) operations. Even if there are duplicates, I’d expect that there are usually significantly fewer duplicates than non-duplicates. So in the grand scheme of things, your change does not improve performance _significantly_, and the worst-case performance stays at O(n²).

It is clear that we will need to check every for every list element whether it is a duplicate. So we cannot get better than O(n). However, we should be able to improve the performance of the lookup whether an element is already in the list. For example, hash tables like `HashMap` and `HashSet` have an expected lookup performance of O(1), which is, of course, much better than doing an O(n) linear search.

So what we are looking for is a solution that uses an efficient lookup data structure (like a hash table). With it, we should be able to achieve an expected worst-case performance of O(n).

I did intentionally not go into more detail, because I do not want to spoil the joy of finding the solution yourself. However, I am happy to elaborate on how I would approach this, if you like.

Reply:
If I understand the idea correctly:

1. Prepare from the list a hash table for lookup.
2. Search a duplicate in the hash table.

@jGleitz  Is it right?

Reply:
Yes, that is how I would approach it.

Reply:
Note that it might be possible to populate the hash table while you are searching, which might be a little more elegant.

Reply:
Thx for the help. I will try.